File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/state/CodeTransformTelemetryState.kt
// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.jetbrains.services.codemodernizer.state

import com.intellij.openapi.components.BaseState
import com.intellij.util.xmlb.annotations.Property
import java.time.Instant
import java.util.UUID

class CodeTransformTelemetryState {
    private var mainState = CodeModernizerTelemetryStateBase()

    fun getSessionId() = mainState.sessionId
    fun setSessionId() {
        mainState.sessionId = UUID.randomUUID().toString()
    }

    fun getStartTime() = mainState.sessionStartTime
    fun setStartTime() {
        mainState.sessionStartTime = Instant.now()
    }

    // Companion object to hold the singleton instance
    companion object {
        // Lazy initialization of the singleton instance
        val instance: CodeTransformTelemetryState by lazy { CodeTransformTelemetryState() }
    }
}

class CodeModernizerTelemetryStateBase : BaseState() {
    @get:Property
    var sessionId: String = UUID.randomUUID().toString()

    @get:Property
    var sessionStartTime: Instant = Instant.now()
}

File: packages/core/src/amazonqGumby/telemetry/codeTransformTelemetryState.ts
/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { randomUUID } from '../../shared/crypto'
import { codeTransformMetaDataToJsonString, ICodeTransformMetaData } from './codeTransformMetadata'
import globals from '../../shared/extensionGlobals'

interface ICodeTransformTelemetryState {
    sessionId: string
    sessionStartTime: number
    resultStatus: string
    codeTransformMetadata: ICodeTransformMetaData
}

export class CodeTransformTelemetryState {
    mainState: ICodeTransformTelemetryState

    private constructor() {
        this.mainState = {
            sessionId: randomUUID(),
            sessionStartTime: globals.clock.Date.now(),
            resultStatus: '',
            codeTransformMetadata: {},
        }
    }

    public getSessionId = () => this.mainState.sessionId
    public getStartTime = () => this.mainState.sessionStartTime
    public getResultStatus = () => this.mainState.resultStatus
    public getCodeTransformMetaData = () => this.mainState.codeTransformMetadata
    public getCodeTransformMetaDataString = () =>
        codeTransformMetaDataToJsonString(this.mainState.codeTransformMetadata)

    public setSessionId = () => {
        this.mainState.sessionId = randomUUID()
    }
    public setStartTime = () => {
        this.mainState.sessionStartTime = globals.clock.Date.now()
    }
    public setResultStatus = (newValue: string) => {
        this.mainState.resultStatus = newValue
    }
    public setCodeTransformMetaDataField = (updatePartial: Partial<ICodeTransformMetaData>) => {
        this.mainState.codeTransformMetadata = {
            ...this.mainState.codeTransformMetadata,
            ...updatePartial,
        }
    }
    public resetCodeTransformMetaDataField = () => (this.mainState.codeTransformMetadata = {})

    static #instance: CodeTransformTelemetryState

    public static get instance() {
        return (this.#instance ??= new this())
    }
}
